blog/content/post/2022-12-16-trenchboot-aem-for-qubesos.md

```diff
@@ -0,0 +1,367 @@
+
+Although the tboot tried to be as OS agnostic as possible, some tboot presence
+awareness from operating system is needed because the application processor
+cores (all cores except the main one) are left in a special state after TXT
+launch and cannot be woken up like in traditional boot process. To solve this
+problem, tboot installs a special processor wakeup procedure in the memory
+which OS must call into to start the processor cores. Only then OS may
+initialize the processor per its own requirements.
+
+As one can see the process is complex in case of Intel TXT. Migration of all
+tboot responsibilities was not trivial and has been divided into the work on
+both GRUB and Xen Hypervisor side of Qubes OS.
+
+## GRUB modifications
+
+![](/img/grub_logo.png)
+
+In order to fulfill the same role as tboot, GRUB had to learn how to prepare
+the platform and perform TXT launch. Most of the work for that particular part
+has been done by [Oracle Team working on TrenchBoot for GRUB](https://www.mail-archive.com/grub-devel@gnu.org/msg30167.html).
+That work however covered the Linux kernel TXT launch only. What still had to
+be done was the multiboot protocol support in GRUB to be able to TXT launch a
+Xen Hypervisor. The patches have been prepared for the respective Qubes GRUB
+[package](https://github.com/3mdeb/qubes-grub2/pull/2).
+
+## Xen modifications
+
+![](/img/xen_project_logo.png)
+
+Analogically to GRUB, Xen had to take over some responsibilities from tboot.
+Due to the Intel TXT requirements for boot process, a new entry point had to be
+developed to which SINIT ACM will return control. The new entry point was
+responsible for saving information that a TXT launch happened and clean up the
+processor state so that the booting of Xen kernel could continue with the
+standard multiboot path. Among others, if Xen detected TXT launch, it had to
+perform the special processor cores wakeup process (which has been rewritten
+from TrenchBoot Linux patches to Xen native code) and measure external
+components before using them (that is the Xen parameters, Dom0 Linux kernel and
+initrd). Xen also had to reserve the memory regions used by Intel TXT, like it
+was done when tboot has been utilized. The relevant source code for the
+respective Qubes Xen package is available [here](https://github.com/3mdeb/qubes-vmm-xen/pull/1).
+
+# Installation and verification of TrenchBoot AEM on Qubes OS
+
+For a seamless deployment and installation of TrenchBoot AEM, the modifications
+have been converted to patches which are applied to projects' sources during
+Qubes OS components compilation. Those patches have been presented earlier with
+links to Pull Requests. It allows building ready-to-use RPM packages that can
+be installed directly on an installed Qubes OS system. Below a procedure for
+building the packages has been presented. If your are not interested in
+compilation, skip to the [next section](#installing-xen-and-grub-packages).
```

Comment:
Installing won't if you don't download (or build) packages first - which user is advised to skip here.

Reply:
Rephrased this section and moved the skipping to the next section

Reply:
Resolved as per: https://github.com/3mdeb/news-and-ideas/pull/503#pullrequestreview-1275863118